### ISSUE TITLE
Help with interpreting the bigstatsr results

### ISSUE THREAD

**Issue by annilk** [privefl/bigstatsr#140]
Hi,

I’m using the bigstatsr function big_spLinReg() to run genome-wide lasso and 5-fold cross-validation on my genotype data. The phenotype is a continuous trait: it is performance in standard episodic memory test called CERAD. As a covariate I use age at the time the memory test scores were measured. The imputed data include 869 samples and 14749154 SNPs. 

My code looks like this:

`covar.train <- covar_from_df(select(ph, "CERADAge"))`
`mod <- big_spLinReg(G, y, covar.train=covar.train, K = 5, ind.sets = ind.sets)`

When running the code I got the warning ‘1991005 variables with low/no variation have been removed’. From the results I can see that summary(mod)$beta includes beta values for 12758150 variants. This seems to make sense since 1991005+12758150=14749154+1 where +1 is for the covariate CERADAge. But how do I know which of the beta values corresponds to this covariate?

I would also need some help with interpreting the plot below. I wonder if the ‘Index’ on the x-axis relates to the lambda values and if yes, then how?

![hbcs_sisu_b38_sanalistat_bigsnpr_5fold_cv](https://user-images.githubusercontent.com/13521152/134331288-a02ff17c-d171-4030-af99-9e00a0925a09.png)

Thank you in advance

**Comment by privefl** [comment]
Have you looked at [the tutorial](https://privefl.github.io/bigstatsr/articles/penalized-regressions.html)?

Also, are you sure you're providing `covar.train` as a string?

***

For your specific questions, look at this code:

```r
library(bigstatsr)

# simulating some data
set.seed(1)
N <- 230
M <- 730
X <- FBM(N, M, init = rnorm(N * M, sd = 5))
X[, 1:5] <- 0
y <- rowSums(X[, 1:10]) + rnorm(N)
covar <- matrix(rnorm(N * 3), N)

# fitting model for multiple lambdas and alphas
mod <- big_spLinReg(X, y, covar.train = covar, K = 3)
```

You should get the warning
```r
Warning message:
5 variables with low/no variation have been removed.
Access remaining columns with 'attr(<object>, "ind.col")'.
```
which means you can use `attr(mod, "ind.col")` to get the `X` indices kept.

For the effects from the covariates, I believe they are placed at the end of this vector.
Therefore, if you have one covariate, you can access the effect with `tail(beta, 1)`, and the effects of the genotypes using `head(beta, -1)`.

***

The index is the index of lambda in the sequence used. Please see the tutorial for more details.
I'll try to update it with more information.

**Comment by annilk** [comment]
Hi,

Thank you for your reply and apologies for the mistake I had made when typing the code in my message. I have now edited the code in the message as it was in my script. In my script, `covar.train` is a matrix including the `CERADAge` values. I have used the `covar_from_df()` to create `covar.train`, as advised in the tutorial. 

And yes, in the warning there was also the line ‘Access remaining columns with 'attr(<object>, "ind.col")‘ but somehow I had missed it. I will use `attr(mod, "ind.col")` to match the SNPs with the beta values.

Thank you also for clarifying the issue with the indices for lambda sequence. 

I would like to know a few more things:

- from the results I can see that the lambda values differ between different folds. Would it possible to run the function ` big_spLinReg()` with fixed lamba values? I tried to search answers for this from the tutorial but didn’t find it. Sorry if I missed it.

- how to derive the R^2 values for each fold? Could they be derived from the validation losses? If yes, I’m a little worried whether it is sensible to take average of the validation loss values over different folds, as the lambda sequence is not the same for each fold. 

- what is the mathematics behind the value given by `summary(mod)$validation_loss`, and how to interpret it? 

Thanks

**Comment by privefl** [comment]
- No, you can't run with specific lambda values. I think CMSA should be preferred over standard cross-validation.

- You can always compute the predicted values for each fold, and compute the R2 yourself. Again, no average is taken as I I'm not implementing cross-validation exactly.

- This is just the standard Gaussian or logistic losses (see https://github.com/privefl/bigstatsr/blob/master/inst/include/bigstatsr/biglasso/linear.hpp#L147 and https://github.com/privefl/bigstatsr/blob/master/inst/include/bigstatsr/biglasso/logistic.hpp#L226).

**Comment by annilk** [comment]
Hi,

Thank you again for your reply. 

I wonder how the lambda values are calculated then? How does CMSA essentially differ from the standard cross-validation?

I have read the description of CMSA given [here](https://privefl.github.io/bigstatsr/reference/big_spLinReg.html) under 'Details'. I have a few questions on the steps taken in CMSA for the function `big_sp_list`:

1. This function separates the training set in K folds (e.g. 10).
> This step I do understand
2.	In turn,

- each fold is considered as an inner validation set and the others (K - 1) folds form an inner training set,

> I understand also this step

- the model is trained on the inner training set and the corresponding predictions (scores) for the inner validation set are computed,

> I guess this step includes finding such regression coefficients that minimize the regularized loss function for the inner training set, and using these coefficients for computing the predictions for the inner validation set?

- the vector of scores which maximizes log-likelihood is determined,

> What is this log-likelihood that is being maximized? 

- the vector of coefficients corresponding to the previous vector of scores is chosen.

3.	The K resulting vectors of coefficients are then averaged into one final vector of coefficients.
> Are some of the steps repeated iteratively until some threshold for e.g. the log-likelihood is achieved?

Thanks again

**Comment by privefl** [comment]
Maximizing the likelihood is equivalent to minimizing the RMSE in the Gaussian case for example.
These are the links to the code I showed before.

Contrary to CV, there is no optimal lambda that is chosen, but instead, K are chosen, one for each train/val. Then, as opposed to CV, there is no retraining of the model with the full training set, instead the K optimal solutions are just averaged.

**Comment by annilk** [comment]
Ok, thank you for your response. 

I'm still a bit confused. When you say that "K optimal solutions are simply averaged" do you mean that average is taken across the K folds, or across the different lambda values in each fold, or something else? What is it exactly that is being averaged? Validation loss? 

I assume that by RMSE you mean root mean squared error - please correct me if I'm wrong. If I'm right, can I calculate R2 as 
`r2 <- 1-summary(mod)$validation_loss^2/var(y)`
?

I'm also a confused about the interpretation of the plot presenting lambda index vs validation loss for each fold. For me it would be intuitive to think that the lower is the validation loss, the better is the model. Therefore, when looking at the plot presented [here](https://privefl.github.io/bigstatsr/articles/penalized-regressions.html), where the validation losses for each fold tend to decrease while lambda index increases, I would think that larger lambda indices generally result in better fit in each fold. But above the plot it says "We can see the validation loss getting worse at some point", which to me sounds as if the decrease of the validation loss would indicate that the fit is getting worse. Could you explain what is that I have misunderstood here?

Thanks again

**Comment by privefl** [comment]
The effects (equivalently the predictions) are averaged.

RMSE if not R2. MSE is just the average of square residuals (errors).

In the tutorial figure (below), the optimal index is ~70 for the purple curve, ~60 for the red one, ~45 for the blue one.
Each optimal lambda corresponds to a model with some effect sizes, which you just average over all folds.
"Validation loss getting worse" actually means getting larger, not lower.

![image](https://user-images.githubusercontent.com/11161036/135995525-2f219d1a-c0bd-41e7-84ef-b5aa1f18e035.png)
